runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -248,33 +253,45 @@ private StateRequestHandler getStateRequestHandler(ExecutableStage executableSta
     return StateRequestHandlers.delegateBasedUponType(handlerMap);
   }
 
-  private static class BagUserStateFactory<K extends ByteString, V, W extends BoundedWindow>
+  static class BagUserStateFactory<K extends ByteString, V, W extends BoundedWindow>
       implements StateRequestHandlers.BagUserStateHandlerFactory<K, V, W> {
 
+    /** Upper limit of the number of valid cache tokens to hand out to the SDK. */
+    private static final int MAX_CACHE_SIZE = 100;
+
     private final StateInternals stateInternals;
     private final KeyedStateBackend<ByteBuffer> keyedStateBackend;
     private final Lock stateBackendLock;
+    /** Cache is scoped by state id. */
```

Comment:
is it possible that there are two DoFns in the pipeline and they accidentally using the same state id? In this case there could be collision of stateid. I think we also need other parameters in the cache key.

Reply:
That is not possible. GreedyPipelineFuser will not fuse two ParDos if they both access user state.

>I think we also need other parameters in the cache key.

Which parameters do we need?